### PR TITLE
KAFKA-16373: KIP-1028:  Adding code to support Apache Kafka Docker Official Images

### PR DESCRIPTION
**Pull request by KrishVora01**
This PR aims to add JVM based Docker Official Image for Apache Kafka as per the following KIP - https://cwiki.apache.org/confluence/display/KAFKA/KIP-1028%3A+Docker+Official+Image+for+Apache+Kafka

This PR adds the following functionalities:
Introduces support for Apache Kafka Docker Official Images via:

1. GitHub Workflows:
      - Workflow to prepare static source files for Docker images
      - Workflow to build and test Docker official images

2. Scripts to prepare source files and perform Docker image builds and tests

3. A new directory for Docker official images, named `docker/docker_official_images`. This is the new directory to house all Docker Official Image assets.

Post this PR:
1. Raise PR with PR template generation script for 3.7.0 under docker official images repo.
2. Add Examples
4. Add Documentation
5. Update post-release process

**Committer Checklist (excluded from commit message)**
- [ ]  Verify design and implementation
- [ ]  Verify test coverage and CI build status
- [ ]  Verify documentation (including upgrade notes)

### CODE REVIEW

File: .github/workflows/docker_official_image_build_and_test.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+on:
+  workflow_dispatch:
+    inputs:
+      image_type:
+        type: choice
+        description: Docker image type to build and test
+        options: 
+          - "jvm"
+      kafka_version:
+        description: Kafka version for which the source for docker official image is to be built
```
Comment (VedarthConfluent) [inline .github/workflows/docker_official_image_build_and_test.yml]: Simplify and mention that version should be >= 3.7.0

File: .github/workflows/prepare_docker_official_image_source.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+on:
+  workflow_dispatch:
+    inputs:
+      image_type:
+        type: choice
+        description: Docker image type to build and test
+        options: 
+          - "jvm"
+      kafka_version:
+        description: Kafka version for which the source for docker official image is to be built
```
Comment (VedarthConfluent) [inline .github/workflows/prepare_docker_official_image_source.yml]: Same as https://github.com/apache/kafka/pull/16027/files#r1611040507

File: .github/workflows/prepare_docker_official_image_source.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    - uses: actions/checkout@v3
+    - name: Set up Python 3.10
+      uses: actions/setup-python@v3
+      with:
+        python-version: "3.10"
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r docker/requirements.txt
+    - name: Build New Artifact
```
Comment (VedarthConfluent) [inline .github/workflows/prepare_docker_official_image_source.yml]: nit: Maybe Build Docker Official Image Source?

File: docker/docker_official_image_build_test.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    kafka_url = f"https://downloads.apache.org/kafka/{args.kafka_version}/kafka_2.13-{args.kafka_version}.tgz"
+    if args.build_only or not (args.build_only or args.test_only):
+        if args.kafka_version:
+            build_docker_official_image(args.image, args.tag, args.kafka_version, args.image_type)
+        else:
+            raise ValueError(
+                "--kafka-version is required argument for jvm docker official image image")
+    if args.test_only or not (args.build_only or args.test_only):
+        run_docker_tests(args.image, args.tag, kafka_url, args.image_type)
+
```
Comment (VedarthConfluent) [inline docker/docker_official_image_build_test.py]: nit: remove extra lines

File: docker/docker_official_images/.gitignore
Comment (VedarthConfluent) [inline docker/docker_official_images/.gitignore:1]: Not needed, we can remove this

Comment (VedarthConfluent) [reply docker/docker_official_images/.gitignore:1]: Can consider using .gitkeep if empty folder is needed

File: docker/generate_kafka_pr_template.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+def get_version_parts(version):
+    return tuple(int(part) for part in version.name.split('.'))
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--image-type", "-type", choices=[
+                        "jvm"], default="jvm", dest="image_type", help="Image type you want to build")
+    args = parser.parse_args()
+    if get_current_branch() != "trunk":
```
Comment (VedarthConfluent) [inline docker/generate_kafka_pr_template.py]: Not needed

File: .github/workflows/docker_official_image_build_and_test.yml (hunk 1 above)
Comment (KrishVora01) [reply .github/workflows/docker_official_image_build_and_test.yml]: Made this change.

File: .github/workflows/prepare_docker_official_image_source.yml (hunk 2 above)
Comment (KrishVora01) [reply .github/workflows/prepare_docker_official_image_source.yml]: Made this change.

File: .github/workflows/prepare_docker_official_image_source.yml (hunk 3 above)
Comment (KrishVora01) [reply .github/workflows/prepare_docker_official_image_source.yml]: Renamed this 👍

File: docker/docker_official_images/.gitignore
Comment (KrishVora01) [reply docker/docker_official_images/.gitignore:1]: modified this to .gitkeep

File: docker/docker_official_image_build_test.py (hunk 4 above)
Comment (KrishVora01) [reply docker/docker_official_image_build_test.py]: removed extra lines from all files

File: docker/generate_kafka_pr_template.py (hunk 5 above)
Comment (KrishVora01) [reply docker/generate_kafka_pr_template.py]: Removed this check

**Comment by omkreddy** [comment]
There is a failure due to missing license:

```
[2024-05-23T07:39:35.270Z] Execution failed for task ':rat'.

[2024-05-23T07:39:35.270Z] > A failure occurred while executing org.nosphere.apache.rat.RatWork

[2024-05-23T07:39:35.270Z]    > Apache Rat audit failure - 1 unapproved license

[2024-05-23T07:39:35.270Z]      	See file:///home/jenkins/workspace/Kafka_kafka-pr_PR-16027/build/rat/index.html
```

**Review by omkreddy — APPROVED** [review]
@KrishVora01 Thanks for the PR. LGTM.

**Pull request merged**